java/sage/io/BufferedSageFile.java

```diff
@@ -0,0 +1,880 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sage.io;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.channels.WritableByteChannel;
+
+public class BufferedSageFile implements SageFileSource
+{
+  // We use this to fill in the write buffer for read only mode.
+  private static final byte[] empty = new byte[0];
+  // 32k is a good default.
+  public static final int READ_BUFFER_SIZE = 32768; // 32k
+  // If this value is greater than the current read buffer size and a byte buffer method is used,
+  // the buffer will be resized up to this value to improve the overall efficiency of transfers.
+  // Otherwise, we will be looping over read, copy, (native) copy a lot more.
+  public static final int READ_BUFFER_NIO_LIMIT = 524288; // 512k
+  // 64k is a good block size on many filesystems. If everything is aligned, we will only be writing
+  // out full blocks unless the way we are randomly writing makes it more optimal to do partial
+  // writes.
+  public static final int WRITE_BUFFER_SIZE = 65536; // 64k
+
+  private final SageFileSource sageFileSource;
+  private final boolean readonly;
+
+  // Cache the last known length so we can possibly avoid getting a value from native which is
+  // usually slower.
+  private long realLastLength = 0;
+  // Keeps track of the actual file position to quickly calculate the apparent position in the file.
+  // Be very mindful of this value and ensure it is always updated correctly or you will have
+  // problems such as reaching EOF early.
+  private long realFilePosition = 0;
+
+  // Used for the transferFrom method, ensure this is made null if the writeBuffer is replaced.
+  private ByteBuffer writeWrap;
+  private byte writeBuffer[];
+  // The write offset is always ahead of the actual file offset by writeOffset.
+  private int writeOffset = 0;
+  // This is used for random writes and seeking to indicate that we are actually ahead or the
+  // current write position for commits. If this value is not 0, we use this instead of writeOffset
+  // to determine how much to write.
+  private int writeLength = 0;
+  // This value is used to determine how many bytes to leave in the write buffer to optimize random
+  // writing. This value is adjusted dynamically based on how far behind the last random write was.
+  private int writeOptimizer = 0;
```

Comment:
Can you explain more what the whole write optimization thing is about? It's somewhat confusing, especially around the usage of writeLength and what it means....but mainly, why it's being done in the first place. I'm assuming it has something to do with an analysis you did regarding how Wizard is writing out files where it always skips backwards to write command lengths and this was done to reduce the number of overall actual disk writes.